### PR TITLE
Add delete_first to Array

### PR DESCRIPTION
**Pull request by dsounded**
`a = ["a", "b", "b", "b", "c"]`
 `a.delete_first("b") => "b"`
 `a # => ["a", "b", "b", "c"]`

 `a = ["a", "b", "b", "b", "c"]`
 `a.delete_first("d") => nil`
 `a # => ["a", "b", "b", "b", "c"]`

### CODE REVIEW

**Comment by jhass** [comment]
I can see the usefulness but I'm not too confident about the naming, it could easily be confused as "except this specific object" or "only this specific object". But then the only alternatives I could come up with, keep and max, are not really better.

File: src/array.cr [hunk 1]
```diff
@@ -671,15 +671,38 @@ class Array(T)
...
   #
   # ```
   # a = ["a", "b", "b", "b", "c"]
   # a.delete("b")
   # a # => ["a", "c"]
   # ```
-  def delete(obj)
-    reject! { |e| e == obj } != nil
+  #
+  # :only option removes only specified number of items from `self` that are equal to *obj*.
```
Comment (jhass) [inline src/array.cr]: They're really not symbols, use `*only*`, see http://crystal-lang.org/docs/conventions/documenting_code.html

Comment (dsounded) [inline src/array.cr]: @jhass ty will check this out

**Comment by asterite** [comment]
I can also see this being useful, but do we need both options? I think having a `max` optional argument should be enough. In which case you want to delete all elements except an arbitrary number of items?

Also, could you show us your use case?

**Comment by dsounded** [comment]
@asterite  It's based on the task: maybe someone will need to take care that some objects will be in the array only needed times(1 for the most cases, but probably the other situation will come as well)

**Comment by asterite** [comment]
@dsounded But do you need it in some project of yours?

I'm hesitant to add methods that don't come from a real use case. In any case, others can reopen Array and add this overload, and eventually if many need it we can consider adding it to the standard library.

**Comment by dsounded** [comment]
I've got similar situation  but with Redis: control count of connection of one user to the other users channels and track if he is active now

**Comment by dsounded** [comment]
@asterite I am trying to move my ws-server from `Rails` to `Kemal(Crystal framework)` and I've got to do that Redis stuff, I can describe it more accurately if you want in a PM(though PR isn't a right place for me).

**Comment by dsounded** [comment]
@asterite can this be merged, or at least `max` option ?

**Comment by asterite** [comment]
@dsounded No, it's not clear that this should be added to the standard library. You can always add such functionality in your projects, either by reopening Array or adding a class method to a module/class.

**Comment by asterite** [comment]
I think instead of this we could add a `delete_first(obj)` or `delete_one(obj)` method. I'm not sure deleting arbitrary n values matching a given object by equality is something common, and it can be programmed as an outer function/method.

**Comment by dsounded** [comment]
@asterite maybe `delete_exactly(obj, count)` or `delete_at_most` or sth like that ?

**Comment by bcardiff** [comment]
I would say that `delete_first(obj)` sounds good to be added to the std. We can add just that.

If we return the index of the deleted object as the result and if we have `delete_first(obj, from)`, implementing an efficient delete-N is trivial. I wouldn't add anything further to the std.

**Comment by dsounded** [comment]
@bcardiff @asterite I've updated it, I am not sure about `from` option, but we can discuss it

**Comment by yxhuvud** [comment]
Isn't this a bit redundant when there is already Array#shift that does the same thing except it also return the first value (which I assume the optimizer is able to throw away)?

**Comment by epergo** [comment]
@dsounded Is anything left to be done in this PR?

It seems it is finished, is a small change so if everyone agrees, I think this could be merged :smiley:

**Comment by dsounded** [comment]
@epergo No, but I am not sure if guys from core team will merge it

**Pull request closed without merging**